modules/boulder/src/main/ts/ephox/boulder/core/ValuePresence.ts

```diff
@@ -0,0 +1,52 @@
+import * as FieldPresence from '../api/FieldPresence';
+import { Processor } from './ValueProcessor';
+
+interface FieldData {
+  readonly key: string;
+  readonly newKey: string;
+  readonly presence: FieldPresence.FieldPresenceTypes;
+  readonly prop: Processor;
+}
+
+interface StateData {
+  readonly newKey: string;
+  readonly instantiator: (obj: any) => unknown;
+}
+
+interface ValuePresenceData<D, T> {
+  readonly discriminator: D;
+  readonly data: T;
+}
+
+export type FieldProcesserData = ValuePresenceData<'field', FieldData>;
+export type StateProcessorData = ValuePresenceData<'state', StateData>;
+
+export type ValueProcessorTypes = FieldProcesserData | StateProcessorData;
```

Comment:
I think sum types should be named in singular form. Since you don't call an option Options for example. If this was an array of ValueProcessorType it could be called ValueProcessorTypes since it contains many of the type ValueProcessorType. You could also use a const enum for the tags that way it compiles to numbers but might make it harder to debug. If possible drop the Type prefix too since it's pretty obvious that it's a type.

Reply:
Good points Spocke, thanks! I'll wait to change things in case anyone else wants to chime in, but happy to run with this unless anyone says otherwise.

Reply:
Maybe `ValueProcessor`?